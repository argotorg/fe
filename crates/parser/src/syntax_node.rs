use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        SyntaxKind::try_from(raw.0).expect("invalid SyntaxKind from raw u16")
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
pub type SyntaxToken = rowan::SyntaxToken<FeLang>;
pub type GreenNode = rowan::GreenNode;
pub type TextRange = rowan::TextRange;
pub type NodeOrToken = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;
