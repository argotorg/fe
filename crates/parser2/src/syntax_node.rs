use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
