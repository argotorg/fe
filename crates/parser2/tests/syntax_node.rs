use fe_parser2::syntax_node::SyntaxNode;

#[allow(unused)]
fn build_cst(input: &str) -> SyntaxNode {
    let (cst, errors) = fe_parser2::parse_source_file(input);
    for error in &errors {
        println!("{}", error.msg);
    }
    assert! {errors.is_empty()}
    cst
}

fe_compiler_test_utils::build_debug_snap_tests! {
    "parser2/test_files/syntax_node",
    "parser2/test_files/syntax_node",
    build_cst
}
