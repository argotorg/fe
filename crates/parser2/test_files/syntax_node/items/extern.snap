---
source: crates/parser2/tests/syntax_node.rs
expression: snapshot
---
Root@0..146
  ItemList@0..146
    Extern@0..11
      ExternKw@0..6 "extern"
      WhiteSpace@6..7 " "
      ExternItemList@7..11
        LBrace@7..8 "{"
        Newline@8..10 "\n\n"
        RBrace@10..11 "}"
    Newline@11..13 "\n\n"
    Extern@13..146
      ExternKw@13..19 "extern"
      WhiteSpace@19..20 " "
      ExternItemList@20..146
        LBrace@20..21 "{"
        Newline@21..22 "\n"
        WhiteSpace@22..26 "    "
        Fn@26..76
          ItemModifier@26..36
            PubKw@26..29 "pub"
            WhiteSpace@29..30 " "
            UnsafeKw@30..36 "unsafe"
          WhiteSpace@36..37 " "
          FnKw@37..39 "fn"
          WhiteSpace@39..40 " "
          Ident@40..45 "write"
          FnArgList@45..68
            LParen@45..46 "("
            FnArg@46..55
              Ident@46..49 "loc"
              Colon@49..50 ":"
              WhiteSpace@50..51 " "
              PtrType@51..55
                Star@51..52 "*"
                PathType@52..55
                  Path@52..55
                    PathSegment@52..55
                      Ident@52..55 "u32"
            Comma@55..56 ","
            WhiteSpace@56..57 " "
            FnArg@57..67
              Ident@57..62 "value"
              Colon@62..63 ":"
              WhiteSpace@63..64 " "
              PathType@64..67
                Path@64..67
                  PathSegment@64..67
                    Ident@64..67 "u32"
            RParen@67..68 ")"
          WhiteSpace@68..69 " "
          Arrow@69..71 "->"
          WhiteSpace@71..72 " "
          PathType@72..76
            Path@72..76
              PathSegment@72..76
                Ident@72..76 "bool"
        Newline@76..77 "\n"
        WhiteSpace@77..81 "    "
        Fn@81..131
          ItemModifier@81..91
            PubKw@81..84 "pub"
            WhiteSpace@84..85 " "
            UnsafeKw@85..91 "unsafe"
          WhiteSpace@91..92 " "
          FnKw@92..94 "fn"
          WhiteSpace@94..95 " "
          Ident@95..99 "read"
          FnArgList@99..122
            LParen@99..100 "("
            FnArg@100..109
              Ident@100..103 "loc"
              Colon@103..104 ":"
              WhiteSpace@104..105 " "
              PtrType@105..109
                Star@105..106 "*"
                PathType@106..109
                  Path@106..109
                    PathSegment@106..109
                      Ident@106..109 "u32"
            Comma@109..110 ","
            WhiteSpace@110..111 " "
            FnArg@111..121
              Ident@111..114 "len"
              Colon@114..115 ":"
              WhiteSpace@115..116 " "
              PathType@116..121
                Path@116..121
                  PathSegment@116..121
                    Ident@116..121 "usize"
            RParen@121..122 ")"
          WhiteSpace@122..123 " "
          Arrow@123..125 "->"
          WhiteSpace@125..126 " "
          PathType@126..131
            Path@126..131
              PathSegment@126..131
                Ident@126..131 "usize"
        Newline@131..132 "\n"
        WhiteSpace@132..136 "    "
        Fn@136..144
          FnKw@136..138 "fn"
          WhiteSpace@138..139 " "
          Ident@139..142 "foo"
          FnArgList@142..144
            LParen@142..143 "("
            RParen@143..144 ")"
        Newline@144..145 "\n"
        RBrace@145..146 "}"

